android/KMEA/app/src/main/java/com/tavultesoft/kmea/KeyboardPickerActivity.java

```diff
@@ -884,16 +714,19 @@ protected static void showLanguageList(Context context) {
   }
 
   public static void handleDownloadedKeyboard(Context context, HashMap<String, String> keyboardInfo) {
-    String keyboardID = keyboardInfo.get(KMManager.KMKey_KeyboardID);
-    String languageID = keyboardInfo.get(KMManager.KMKey_LanguageID);
-    String kbKey = String.format("%s_%s", languageID, keyboardID);
-    int index = getKeyboardIndex(context, kbKey);
-    if (index == -1) {
-      // Add the downloaded keyboard if not found
-      addKeyboard(context, keyboardInfo);
-      index = getKeyboardIndex(context, kbKey);
-    }
-    keyboardsList.set(index, keyboardInfo);
-    saveList(context, KMManager.KMFilename_KeyboardsList);
+    boolean isNewKeyboard = true;
+    Keyboard k = new Keyboard(
+      keyboardInfo.get(KMManager.KMKey_PackageID),
+      keyboardInfo.get(KMManager.KMKey_KeyboardID),
+      keyboardInfo.get(KMManager.KMKey_KeyboardName),
+      keyboardInfo.get(KMManager.KMKey_LanguageID),
+      keyboardInfo.get(KMManager.KMKey_LanguageName),
+      keyboardInfo.get(KMManager.KMKey_KeyboardVersion),
+      keyboardInfo.get(KMManager.KMKey_HelpLink),
+      isNewKeyboard,
+      keyboardInfo.get(KMManager.KMKey_Font),
+      keyboardInfo.get(KMManager.KMKey_OskFont));
+    KeyboardController.getInstance().add(k);
```

Comment:
Might be some chance for further improvement here; why is this method not called with a pre-constructed `Keyboard`?

Reply:
This is called by `ResourceUpdateTool.onKeyboardDownloadFinished` which still uses HashMaps (more than I want to tackle on this PR).